server/devpi_server/keyfs_sqlite_fs.py

```diff
@@ -36,10 +36,18 @@ def __init__(self, path):
             digest = digest[:8]
         self.tmppath = f"{path}-{digest}-tmp"
 
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self.path})"
+
     @classmethod
     def from_content(cls, path, content_or_file):
         self = DirtyFile(path)
-        if hasattr(content_or_file, "devpi_srcpath"):
+        if hasattr(content_or_file, "filestore"):
```

Comment:
Instead of passing in a link, you should create a new link and pass in the file with the ``devpi_srcpath`` attribute set, that way it works with all storage backends like ``devpi-postgresql`` which doesn't have files on the filesystem and currently doesn't support hard links (but it could with a schema change).